lucene/core/src/java/org/apache/lucene/search/LeafCollector.java

```diff
@@ -93,4 +93,16 @@
    */
   void collect(int doc) throws IOException;
 
+  /**
+   * Optionally creates a view of the scorerIterator where only competitive documents
+   * in the scorerIterator are kept and non-competitive are skipped.
+   *
+   * Collectors should delegate this method to their comparators if
+   * their comparators provide the skipping functionality over non-competitive docs.
+   * The default is to return the same iterator which is interpreted as the collector doesn't filter any documents.
+   */
+  default DocIdSetIterator filterIterator(DocIdSetIterator scorerIterator) {
+    return scorerIterator;
+  }
```

Comment:
This allows for some hacks like returning an iterator that matches more hacks than the scorer. I liked the previous approach that returned an iterator better.

Reply:
Oh that's a good point. +1 to just return an iterator based on the comparator, and do the conjuncion/combination in `BulkScorer`

Reply:
Thanks, makes sense. Addressed in d7ef9b6